### PR TITLE
do not mount /dev/shm or /dev/mqueue if we are mounting from the host

### PR DESCRIPTION
**Pull request by jessfraz**
thinking about a test...

### CODE REVIEW

**Comment by jfrazelle** [comment]
also... food for thought... should this even be allowed how i was mounting w -v, when that is now how `--ipc host` works

**Comment by jfrazelle** [comment]
butttt it did work before so idk how many people like me rely on it...

**Comment by tiborvass** [comment]
Related to #15862

**Comment by tiborvass** [comment]
Ping @mrunalp @estesp

File: daemon/container_unix.go [hunk 1]
```diff
@@ -1305,20 +1305,40 @@ func (container *Container) unmountIpcMounts() error {
...
-		Destination: "/dev/mqueue",
-		Writable:    true,
-		Private:     true,
-	})
+
+	// check if mounting shm or mqueue from host
+	var shmFromHost, mqueueFromHost bool
+	for _, mount := range container.MountPoints {
+		if mount.Destination == "/dev/shm" {
+			shmFromHost = true
```
Comment (tiborvass) [inline daemon/container_unix.go]: `shmFromHost = mount.Destination == "/dev/shm"` ?

**Comment by mrunalp** [comment]
This seems reasonable. You also want to make sure there no cleanup happens in this case i.e. no attempt to unmount and remove the shm/mqueue.

**Comment by jfrazelle** [comment]
ah yes fixing cleanup

**Comment by cpuguy83** [comment]
daemon/daemon_unix.go:539: unknown Container field 'MountPoints' in struct literal

**Comment by jfrazelle** [comment]
I know a test, I'll update :p

**Comment by jfrazelle** [comment]
updated with a test \o/

**Comment by jfrazelle** [comment]
it fails on master :p

**Comment by vdemeester** [comment]
> updated with a test \o/

giphy_minions1

**Comment by cpuguy83** [comment]
Instead of moving Mountpoints around, might be better to handle this in the same way as the network mounts are handled.
https://github.com/docker/docker/blob/master/daemon/volumes_unix.go#L48

**Comment by jfrazelle** [comment]
I don't know if I like this better or if this was what you had in mind cc @cpuguy83

**Comment by jfrazelle** [comment]
windows is failing because of something unrelated, ive seen it on other PRs too

**Comment by estesp** [comment]
Latest version looks good/cleaner.  **LGTM**

**Comment by jfrazelle** [comment]
all thanks to @cpuguy83 :)

**Comment by cpuguy83** [comment]
LGTM

**Pull request merged**